### PR TITLE
Freeze header row functionality

### PR DESCRIPTION
**Pull request by quadrophobiac**
opening a PR to see if I am on the right path with this or in case I'm missing an easy trick.
I have made the minimum changes required to freeze a row for scrolling purposes. I've left it open for a user to freeze multiple rows even though a CSV headers approach would prefer for this not to be the case.
As far as I can tell Electron doesn't support dynamics changes to menu items which would fit with a toggled 'freeze row / unfreeze row' addition to both menu items.
@chris48s I'm blocked on the inclusion of these functions in the `main` process menu. Basically it seems like nothing is firing between `main` and `renderer`. Was wondering if you could take a glance at the code and tell me if I am doing anything glaringly wrong or if there is a better way of writing code the communicates between the main and renderer process in Electron nowadays

### CODE REVIEW

**Comment by quadrophobiac** [comment]
for fix to #148 sans the header-dialect functionality

**Comment by Stephen-Gates** [comment]
Should be able to toggle menu value. Implemented in Atom 
screenshot 2017-06-26 23 00 34

**Comment by chris48s** [comment]
Really quick off-the-cuff response. I've not checked out the code and run it, but I think the piece of the jigsaw you are missing is that you haven't defined handlers for the `freezeRows` and `unfreeze` events.

If you add a few lines to https://github.com/theodi/comma-chameleon/blob/master/renderer/index.js declaring handlers like

```
ipc.on('freezeRows', function() {
  // do the stuff
});
```
..that should get you moving :)

**Comment by quadrophobiac** [comment]
Thank you for the pointer @chris48s , still reacclimatising to Electron environment, that's unblocked it!

**Comment by quadrophobiac** [comment]
This is basically the bare bones functionality to freeze the header. What could be improved in subsequent patches are...

- [ ] colour coding of rows frozen
- [ ] consolidation of freeze and unfreeze functions into a single function that toggle in the relevant Electron menus
- [ ] CSV Dialect Support indicated in #148 

Additions suffice to add mvp user functionality

**Comment by quadrophobiac** [comment]
@chris48s I made the suggestions, with more detail [here](https://github.com/theodi/comma-chameleon/pull/169/commits/d1026cec2c4e26da5b90eb616c7cd3fd2a54527f). 
To get the row to freeze consistently I had to kludge it though - calling `updateSettings` twice. I'm not sure if that's necessary or if my understanding of javascript is leading to a very poor solution. However when I tested the callbacks [via script from here https://docs.handsontable.com/0.16.1/tutorial-callbacks.html] I could see that the re-rendering was executing prior to the `fixedRowsTop` value being altered. I'm inclined to think that this might be a bug that could be fixed in subsequent versions of HandsOnTable but even still I'm not certain about incorporating the kuldge to codebase - thoughts?

**Comment by chris48s** [comment]
I've got several bits of time-consuming life admin going on at the moment so I'm going to aim to look at this on Monday evening next week for you. If that's blocking you, happy to defer to @pezholio or @Floppy 's reckons :)

File: renderer/hot.js [hunk 1]
```diff
@@ -4,42 +4,43 @@ var loader = require('../renderer/loader.js');
...
+              }
+          }
+          return {row: 0, col: 1};
+      },
+      afterInit: function() {
+          loader.showLoader('Loading...');
+      },
+      afterLoadData: function() {
+          loader.hideLoader();
+      },
```
Comment (chris48s) [inline renderer/hot.js]: Add a handler for the `afterUpdateSettings` event here

File: renderer/index.js [hunk 2]
```diff
@@ -153,3 +153,10 @@ ipc.on('removeRows', function() {
 ipc.on('removeColumns', function() {
   hotController.removeColumns();
 });
+
+ipc.on('freezeRows', function(){
```
Comment (chris48s) [inline renderer/index.js]: This event hander should be updated to expect an event called `'freeze'` to match with the event you're raising [here](https://github.com/theodi/comma-chameleon/pull/169/files?w=1#diff-b8bbf1b64cab439cc1c4ead74d50d1aeR203)

File: renderer/menu.js [hunk 3]
```diff
@@ -47,6 +47,23 @@ var removeCol = new MenuItem({
...
 
+var freezeRow = new MenuItem({
+    label: 'Freeze Header Row',
+    click: function(){
+      hotController.freeze();
+    }
+});
+
+var unfreezeRow = new MenuItem({
+    label: 'unfreeze header row',
```
Comment (chris48s) [inline renderer/menu.js]: 'Unfreeze Header Row'

**Comment by chris48s** [comment]
Managed to slot in a bit of time to look at this tonight for you. All the odd behaviours I reported in my previous post seem to be fixed now, so that's useful.

There's one minor problem: The right-click menu item works but the 'Freeze Header Row' in the Edit menu item doesn't. It is firing an event called `freeze` but the registered handler is expecting an event called `freezeRows`. I've left a line note where it needs to be changed.


> To get the row to freeze consistently I had to kludge it though - calling `updateSettings` twice. I'm not sure if that's necessary or if my understanding of javascript is leading to a very poor solution

Yeah - I can see the issue you were having. I'm also not entirely sure if this is intended behaviour for HoT or a bug we need to work around, but I think there is a neater way to ensure the re-rendering occurs after the setting are applied. Define a `afterUpdateSettings` handler when we pass the initial config to the `Handsontable` constructor, like this:

```
...
afterUpdateSettings: function() {
  hot.render();
  hot.deselectCell();
},
...
```

I've also added a line note where that declaration should live. Then that allows you to simplify your event handlers to:

```
var unfreezeHeaderRow = function(){
    hot.updateSettings({fixedRowsTop: 0, colHeaders: true});
};

var freezeHeaderRow = function(){
    hot.updateSettings({fixedRowsTop: 1}, false);
};
```

and deal with the cleanup in the handler function (which is guaranteed to execute after the settings have been applied). Maybe also note in issue #159 that we might be able to remove that manual call to `render()` when we upgrade the library.

**Comment by quadrophobiac** [comment]
Hey @chris48s , thanks so much for taking the time to give me solid js pointers on how to achieve the requisite functionality in a more cohesive way really appreciate the tuition you're imparting! And aside from personal gratitude I'd also like to thank you because having your insight on hand really helps the work that can be done on Comma Chameleon during present circumstances - so thanks from the Toolbox too :)

**Comment by chris48s** [comment]
no probs - glad to help :)

**Comment by quadrophobiac** [comment]
Cool, so I'm taking your input as a code review @chris48s and I'm going to merge this in now

**Pull request merged**